README
Geo-Address-Mail

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Geo::Address::Mail

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Geo-Address-Mail

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Geo-Address-Mail

    CPAN Ratings
        http://cpanratings.perl.org/d/Geo-Address-Mail

    Search CPAN
        http://search.cpan.org/dist/Geo-Address-Mail/


COPYRIGHT AND LICENCE

Copyright (C) 2010 Cory G Watson

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Geo-Address-Mail

0.03    Mon Jan 3, 2010
  * Add MooseX::Storage depency (thanks Andreas Koenig)

0.02    Tue Dec 28, 2010
  * Add MooseX::Storage (holophrastic)

0.01    Date/time
        First version, released on an unsuspecting world.



lib/Geo/Address/Mail/US.pm
package Geo::Address::Mail::US;
use warnings;
use strict;
use Moose;

use Moose::Util::TypeConstraints;
use Regexp::Common qw(zip);

extends 'Geo::Address::Mail';

subtype 'Geo::Address::Mail::USPostalCode',
    => as 'Str',
    => where { $_ =~ /^$RE{zip}{US}$/ };

has '+postal_code' => (
    isa => 'Geo::Address::Mail::USPostalCode'
);

has 'street2' => (
    is => 'rw',
    isa => 'Str'
);

has 'state' => (
    is => 'rw',
    isa => 'Str'
);

__PACKAGE__->meta->make_immutable;
no Moose;
1;

=head1 NAME

Geo::Address::Mail::US - A Mailing Address in the United States

=head1 SYNOPSIS

Geo::Address::Mail::US is a subclass of L<Geo::Address::Mail> that provides
specific validation and attributes for mailing addresses located in the United
States.

    use Geo::Address::Mail::US;

    my $add = Geo::Address::Mail::US->new(
        name => 'Cory G Watson',
        street => '123 Main St',
        street2 => 'Apt 3B',
        city => 'Testville',
        postal_code => '12345'
    );

=head1 ATTRIBUTES

Geo::Address::Mail::US has all the attributes of L<Geo::Address::Mail>.  The
following attributes are either modified or new.

=head2 postal_code

Postal codes are validated to conform to the USPS ZIP (and optional +4)
standard.

=head2 street2

Addresses in the United States often have an Apartment number, Suite number
or other sub-street unit.  This field provides for that.

=head2 state

The state for this address. Added in US as not all countries have the
concept of a state.

=head1 AUTHOR

Cory G Watson, C<< <gphat at cpan.org> >>

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2010 Cory G Watson.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


lib/Geo/Address/Mail.pm
package Geo::Address::Mail;
use warnings;
use strict;
use Moose;
use MooseX::Storage;
with qw(MooseX::Storage::Deferred);
with qw(MooseX::Clone);

our $VERSION = '0.03';

has 'city' => (
    is => 'rw',
    isa => 'Str'
);

has 'company' => (
    is => 'rw',
    isa => 'Str'
);

has 'country' => (
    is => 'rw',
    isa => 'Str'
);

has 'name' => (
    is => 'rw',
    isa => 'Str'
);

has 'postal_code' => (
    is => 'rw',
    isa => 'Str',
);

has 'street' => (
    is => 'rw',
    isa => 'Str'
);

__PACKAGE__->meta->make_immutable;
no Moose;
1;

=head1 NAME

Geo::Address::Mail - A Mailing Address on Earth

=head1 SYNOPSIS

Geo::Address::Mail provides a generic object representation of a mailing
address that may be subclassed to provide more specific typing of attributes.
The core class, Geo::Address::Mail provides common, loosely typed attributes
and methods.

    use Geo::Address::Mail::US;

    my $add = Geo::Address::Mail::US->new(
        name => 'Cory G Watson',
        street => '123 Main St',
        city => 'Testville',
        postal_code => '12345'
    );

=head1 SUBCLASSING

The real reason for Geo::Address::Mail is to provide a common class that can
be used to build mailing address objects for other countries.

Subclasses are expected additional type refinement and attributes.
For example, L<Geo::Address::Mail::US> uses a more specific type for
validation USPS ZIP codes and adds a C<street2> attribute for an optional
additional line of addressing.

=head1 ADDITIONAL USES

Using a common address object enables a family of distributions that provide
interesting address functionality such as L<Geo::Address::Mail::Standardizer>.

=head1 ATTRIBUTES

=head2 city

The city/town/village/municipality in which this address resides.

=head2 company

The name of the company that is to receive the mail piece.

=head2 country

The country in which this address resides.  This is likely not necessary
unless the address is in a different country.

=head2 name

The name of the person that is to receive the mail piece.

=head2 postal_code

The postal code of the address.  Called the ZIP code in the US.

=head2 street

The number and name of the street that is to receive the mail piece.

  2020 Main St

=head1 AUTHOR

Cory G Watson, C<< <gphat at cpan.org> >>

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2010 Cory G Watson.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut
